### ISSUE TITLE
[BUG] errors compiling the hello world exmaple

### ISSUE THREAD

**Issue by hros** [hsutter/cppfront#1334]
**Describe the bug**
I get a couple of compiler errors compiling `hello.cpp` which is generated from the `hello.cpp2` example

**To Reproduce**
1. Running `cppfront` on `hello.cpp2` from the [documentation](https://hsutter.github.io/cppfront/welcome/hello-world/) results in
<details>
<summary>hello.cpp</summary>
#define CPP2_IMPORT_STD          Yes

//=== Cpp2 type declarations ====================================================
#include "cpp2util.h"

#line 1 "hello.cpp2"

//=== Cpp2 type definitions and function declarations ===========================

#line 1 "hello.cpp2"
auto main() -> int;

#line 7 "hello.cpp2"
[[nodiscard]] auto hello(cpp2::impl::in<std::string_view> msg) -> decltype(auto);

//=== Cpp2 function definitions =================================================

#line 1 "hello.cpp2"
auto main() -> int{
#line 2 "hello.cpp2"
    std::vector words {"Alice", "Bob"};
    hello(CPP2_ASSERT_IN_BOUNDS_LITERAL(words, 0));
    hello(CPP2_ASSERT_IN_BOUNDS_LITERAL(cpp2::move(words), 1));
}

#line 7 "hello.cpp2"
[[nodiscard]] auto hello(cpp2::impl::in<std::string_view> msg) -> decltype(auto) {
    return std::cout << "Hello, " + cpp2::to_string(msg) + "!\n";  }
</details>

2. Command lines including which C++ compiler you are using: 
g++ version 14.2.0 on Ubuntu 24.4
`g++ -std=c++20 hello.cpp -o hello -I$HOME/.local/cppfront/include`

3. Expected result - compile successfully

4. Actual result/error:

```log
hello.cpp2: In function ‘int main()’:
hello.cpp2:3:10: error: use of ‘decltype(auto) hello(cpp2::impl::in<std::basic_string_view<char> >)’ before deduction of ‘auto’
    3 |     hello( words[0] );
      |     ~~~~~^~~~~~~~~~~~~
hello.cpp2:4:10: error: use of ‘decltype(auto) hello(cpp2::impl::in<std::basic_string_view<char> >)’ before deduction of ‘auto’
    4 |     hello( words[1] );
      |     ~~~~~^~~~~~~~~~~~~
```
**Additional context**

**Comment by gregmarr** [comment]
The clang version: 

cpp2:
```
hello: (msg: std::string_view) =
    std::cout << "Hello, (msg)$!\n";
```

cpp1:
```
auto hello(std::string_view msg) -> decltype(auto);
    
auto main() -> int{
    std::vector words {"Alice", "Bob"}; 
    hello(words[0]);
    hello(words[1]);
}

auto hello(std::string_view msg) -> decltype(auto) { 
    return std::cout << "Hello, " + std::string(msg) + "!\n";
}
```

clang error:
```
error: function 'hello' with deduced return type cannot be used before it is defined
   11 |     hello(words[0]);
      |     ^
```

**Comment by DyXel** [comment]
This became order-dependent after the tersest lambda change. Adding `-> void` to `hello` should fix it.

**Comment by gregmarr** [comment]
If it's order dependent, should cppfront still be writing out the declaration before the definition?

**Comment by DyXel** [comment]
Good point, not sure. I think the current error is better than simply getting a "undeclared identifier". At least it gives you a hint about what the problem might be.

**Comment by gregmarr** [comment]
The Clang error is good, MSVC is also good:
```
<source>(5): error C3779: 'hello': a function that returns 'decltype(auto)' cannot be used before it is defined
<source>(3): note: see declaration of 'hello'
```

The GCC error could use some clarification.  Once you know what the error means, you can squint and see the meaning there, but it's not immediately obvious.

```
use of ‘decltype(auto) hello(cpp2::impl::in<std::basic_string_view<char> >)’ before deduction of ‘auto’
```

Without the forward declaration, the errors are: 

```
GCC: error: 'hello' was not declared in this scope
Clang: error: use of undeclared identifier 'hello'
MSVC: error C3861: 'hello': identifier not found
```

So it's a tossup whether it's better or not.  I'm guessing most C++ developers are familiar with order dependence, and know what these errors mean.  

Maybe cppfront can detect the order-dependent calls and produce an error?  I suppose that would be difficult, since there could be another version that was `#include`d from somewhere that it doesn't know about.

**Comment by ivo-doko** [comment]
> This became order-dependent after the tersest lambda change. Adding `-> void` to `hello` should fix it.

This seems like an issue in itself.

[Order independence is one of the design stakes of Cpp2/cppfront](https://youtu.be/ELeZAKCN4tY?t=1061) - "No forward declarations or ordering gotchas."

If it has turned out that this is not doable, then cppfront should accept and embrace the order-dependent nature it is forced to have, rather than pretend-having order independence, since that just ends up making issues caused by order dependence easier to stumble into and harder to figure out.

**Comment by gregmarr** [comment]
>If it has turned out that this is not doable,

It is not available *right now for free functions with deduced return types*.  It's too early to say that it's not doable at all for *free functions with deduced return types*.  Even if that were to be true, that just means that *free functions with deduced return types* are not order-independent.  Functions with fixed return types still are.  It also shouldn't affect member functions.  Whether that is a large percentage of the code or not still remains to be seen.

**Comment by LT2H** [comment]
Adding `{ }` to the `hello` function also fixes this.

```
main: () = {
    words: std::vector = ( "Alice", "Bob" );
    hello( words[0] );
    hello( words[1] );
}

hello: (msg: std::string_view) = {
    std::cout << "Hello, (msg)$!\n";
}
```
hello.cpp
```
auto main() -> int;

auto hello(cpp2::impl::in<std::string_view> msg) -> void;
auto main() -> int{
    std::vector words {"Alice", "Bob"}; 
    hello(CPP2_ASSERT_IN_BOUNDS_LITERAL(words, 0));
    hello(CPP2_ASSERT_IN_BOUNDS_LITERAL(cpp2::move(words), 1));
}

auto hello(cpp2::impl::in<std::string_view> msg) -> void{
    std::cout << "Hello, " + cpp2::to_string(msg) + "!\n";
}
```

**Comment by ivo-doko** [comment]
> Adding `{ }` to the `hello` function also fixes this.

That seems odd.

That makes me think that this should be resolvable with an appropriate parsing/translation improvement - omitting `{ }` from function definitions ought to be merely a syntactic-sugar-level convenience, so I don't see a reason for that to have this sort of effect.

**Comment by gregmarr** [comment]
There is discussion of this exact thing in the notes on the last release: 
https://github.com/hsutter/cppfront/releases/tag/v0.8.0

> an error that a deduced return type makes the function order-dependent, when the function is called from earlier in the source file... this is because a deduced return types creates a dependency on the body, and it is inherent (not an artifact of either Cpp1 or Cpp2)

**Comment by ivo-doko** [comment]
> Omitting { } translates to _returning_ the expression.

> you're not just omitting the `{}` from the function.

D'oh, apologies, I knew that, but just forgot.

**Comment by vovaskochko** [comment]
Any updates on this thread? I just go through the documentation and it seems a bit confusing to have compilation error on the simplest program. Maybe some comment should be added to documentation to save some time for people who are not aware of the issue.

**Comment by gregmarr** [comment]
Herb stated recently that he had to take some time away from CppFront to focus on C++ standardization work, so issues and pull requests here would need to take a back seat.

**Comment by vovaskochko** [comment]
Yeah - I read that message from Herb.

@gregmarr Thank you for the patch.
